dsl_parser/functions.py

```diff
@@ -772,115 +734,94 @@ def parse(raw_function, scope=None, context=None, path=None):
     return payload
 
 
-def evaluate_capabilities(capabilities,
-                          get_node_instances_method,
-                          get_node_instance_method,
-                          get_node_method,
-                          get_secret_method,
-                          get_capability_method):
+def evaluate_capabilities(capabilities, storage):
     """Evaluates capabilities definition containing intrinsic functions.
 
     :param capabilities: The dict of capabilities to evaluate
-    :param get_node_instances_method: A method for getting node instances.
-    :param get_node_instance_method: A method for getting a node instance.
-    :param get_node_method: A method for getting a node.
-    :param get_secret_method: A method for getting a secret.
-    :param get_capability_method: A method for getting a capability.
+    :param storage: Storage backend for runtime function evaluation
     :return: Capabilities dict.
     """
     capabilities = dict((k, v['value']) for k, v in capabilities.items())
     return evaluate_functions(
         payload=capabilities,
         context={},
-        get_node_instances_method=get_node_instances_method,
-        get_node_instance_method=get_node_instance_method,
-        get_node_method=get_node_method,
-        get_secret_method=get_secret_method,
-        get_capability_method=get_capability_method)
-
-
-def evaluate_outputs(outputs_def,
-                     get_node_instances_method,
-                     get_node_instance_method,
-                     get_node_method,
-                     get_secret_method,
-                     get_capability_method):
+        storage=storage)
+
+
+def evaluate_outputs(outputs_def, storage):
     """Evaluates an outputs definition containing intrinsic functions.
 
     :param outputs_def: Outputs definition.
-    :param get_node_instances_method: A method for getting node instances.
-    :param get_node_instance_method: A method for getting a node instance.
-    :param get_node_method: A method for getting a node.
-    :param get_secret_method: A method for getting a secret.
-    :param get_capability_method: A method for getting a capability.
+    :param storage: Storage backend for runtime function evaluation
     :return: Outputs dict.
     """
     outputs = dict((k, v['value']) for k, v in outputs_def.iteritems())
     return evaluate_functions(
         payload=outputs,
         context={'evaluate_outputs': True},
-        get_node_instances_method=get_node_instances_method,
-        get_node_instance_method=get_node_instance_method,
-        get_node_method=get_node_method,
-        get_secret_method=get_secret_method,
-        get_capability_method=get_capability_method)
+        storage=storage)
 
 
-def _handler(evaluator, **evaluator_kwargs):
-    def _args_to_str_func(args, kwargs):
-        """Used to display extra information when recursion limit is reached.
+def _args_to_str_func(handler, v, scope, context, path):
```

Comment:
This doesn't match the signature you said in the RecursionLimit class where you announced it should be fun(args, kwargs)